doc: Add `x86_64-w64-mingw32ucrt` triplet to `depends/README.md`

hebasto opened the pull request:
This PR is part of the ongoing effort to migrate to the modern UCRT runtime for cross-compiled Windows binaries, including release builds.

For more details about this migration, see:
- https://github.com/bitcoin/bitcoin/issues/30210
- https://github.com/bitcoin/bitcoin/pull/33593
- https://github.com/bitcoin/bitcoin/pull/33764

Can be tested on the following systems:
-  Debian Trixie x86_64 (requires the [`g++-mingw-w64-ucrt64`](https://packages.debian.org/trixie/g++-mingw-w64-ucrt64) package, as documented).
-  Fedora 42 or 43 (requires the [`ucrt64-gcc-c++`](https://packages.fedoraproject.org/pkgs/mingw-gcc/ucrt64-gcc-c++/) package).

Also see related upstream issues:
- https://bugs.launchpad.net/ubuntu/+source/gcc-mingw-w64/+bug/2132114
- https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1121403

hebasto commented:
> Can be tested on the following systems:
> 
>    * Fedora 42 or 43 (requires the `ucrt64-gcc-c++` package).

For example: https://github.com/hebasto/bitcoin-core-nightly/actions/runs/19274168370/job/55109785216.

fanquake commented on depends/README.md:
```diff
@@ -122,7 +122,8 @@ Common `host-platform-triplet`s for cross compilation are:
 - `i686-pc-linux-gnu` for Linux x86 32 bit
 - `x86_64-pc-linux-gnu` for Linux x86 64 bit
-- `x86_64-w64-mingw32` for Win64
+- `x86_64-w64-mingw32` for Windows using MSVCRT
```
How will a user know what to pick (here and below), given we don't explain the difference? Should one be marked as deprecated/unsupported? Probably easier to wait until we only support one, and then we don't need to worry about this.

hebasto replied:
Given that full UCRT-capable toolchains are not yet available across all major distributions, developers on Ubuntu, for example, may still want to build for `x86_64-w64-mingw32`.

hebasto replied:
> Ah ok, so maybe someone could open a question/request at https://launchpad.net/ubuntu/+source/gcc-mingw-w64/+bugs, so that the progress can be tracked.

https://bugs.launchpad.net/ubuntu/+source/gcc-mingw-w64/+bug/2132114

hebasto replied:
> > developers on Ubuntu, for example, may still want to build for x86_64-w64-mingw32.
> 
> Are we supporting that, if the CI is removed in #33593 (along with dropping workarounds going forward).

We won't support that once workarounds are dropped and code become incompatible with MSVCRT. By that point, I don’t anticipate any issues with providing that support.

fanquake replied:
> By that point, 

Isn't that point #33593? The workarounds should be removed along with the CI being dropped, otherwise they are untested and unused.

hebasto replied:
> Isn't that point #33593? The workarounds should be removed along with the CI being dropped, otherwise they are untested and unused.

It could be in #33593 or in a follow-up PR.

fanquake reviewed (commented):
If we are migrating, should we just do this at this same time as we actually migrate, then we don't have to worry about both. Are we dropping support for the old runtime at the same time we switch to using the new ones in releases?

hebasto commented on depends/README.md:
```diff
@@ -137,16 +138,20 @@ The paths are automatically configured and no other options are needed.
 #### For macOS cross compilation
 
-    apt install clang lld llvm zip
+    sudo apt install clang lld llvm zip
```
Thanks! Dropped.

hebasto commented:
> > In my view, dropping support for x86_64-w64-mingw32 should be considered separately and at a later stage, once the codebase becomes incompatible with MSVCRT or the toolchain landscape has matured.
> 
> Shouldn't it be removed once we remove it from the CI? which, according to the PR description in #33764, will be done once the Guix migration is done:

Agreed.

> > MSVCRT-related CI jobs should be removed from the CI framework once the migration to UCRT is complete.
> 
> Just trying to get a clearer idea of what order things are happening in, and what we are supporting and for how long.

I don't think the precise coordination between PRs addressing the subtasks outlined in https://github.com/bitcoin/bitcoin/issues/30210 is essential.

m3dwards commented:
> I’m unable to reproduce this in a fresh Trixie container using Docker

I did almost the same apart from I checked out this PR rather than building from master.

This works on x86 but isn't working on Apple ARM for me. I don't think it's anything wrong with this PR, but it's a problem with the UCRT MinGW compiler. That being the case, should this hold up this PR?

hebasto commented:
> > I’m unable to reproduce this in a fresh Trixie container using Docker
> 
> I did almost the same apart from I checked out this PR rather than building from master.

Sorry. I've adjusted my comment.

> This works on x86 but isn't working on Apple ARM for me.

I can confirm the issue on `aarch64`.

m3dwards commented:
> I can confirm the issue on `aarch64`

According to the Debian bug report, seems it's also an issue on RISC-V.

hebasto committed:
ci: Remove redundant `DEP_OPTS` from “Windows-cross UCRT” job

GCC-based tools already follow the standard naming convention for the
`x86_64-w64-mingw32ucrt` target.

hebasto committed:
doc: Add `x86_64-w64-mingw32ucrt` triplet to `depends/README.md`

hebasto commented:
The feedback from @m3dwards has been addressed.

hebasto commented on depends/hosts/mingw32.mk:
```diff
@@ -1,9 +1,11 @@
+ifneq (ucrt,$(findstring ucrt,$(host)))
```
Thanks! [Reworked](https://github.com/bitcoin/bitcoin/pull/33857#issuecomment-3592479046).

hodlinator reviewed (approved):
ACK ec8eb013a9bfceb324b309f13b8946b05292a993

Ran
`env -i HOME="$HOME" PATH="$PATH" USER="$USER" bash -c 'MAKEJOBS="-j$(nproc)" FILE_ENV="./ci/test/00_setup_env_win64.sh" ./ci/test_run_all.sh'`
locally and confirmed it successfully generated an executable at
`/ci_container_base/ci/scratch/build-x86_64-w64-mingw32ucrt/bin/bitcoind.exe`
